src/StrawberryShake/Client/src/Transport.WebSockets/Protocols/GraphQlWs/SubscriptionTransportWsProtocol.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using StrawberryShake.Http.Subscriptions;
+using StrawberryShake.Transport.Http;
+
+namespace StrawberryShake.Transport.WebSockets
+{
+    public class SubscriptionTransportWsProtocol : ISocketProtocol
+    {
+        private readonly JsonOperationRequestSerializer _serializer = new();
+        private readonly MessagePipelineHandler _handler;
+        private readonly HashSet<OnReceiveAsync> _listeners = new();
+
+        private ISocketClient? _socketClient;
+        private bool _disposed;
+
+        public SubscriptionTransportWsProtocol()
+        {
+            _handler = new MessagePipelineHandler(this);
```

Comment:
On the protocol we could register specific handlers for messaged. e.g. a connection init handler